File: tests/gpu/eco_residue_test.py
"""THE ECONOMY RESIDUE — the GPU half.

    python tests/gpu/eco_residue_test.py

The TS twin is tests/cpu/city/eco-residue.test.ts.

  1. a building buys off its fractional scaled cost (`_b_cols` "buyCost"): the
     Granary (Cost 65) for 130 gold, where a unit keeps its truncated cost.
  2. the policy unlock (`_policy_unlock_cost`): free the turn after a civic,
     then 50 dropping 5 a turn to 10; a government change outside the window
     pays it once, and one the purse cannot meet is refused.
  3. a city-state's city holds the Palace (`_palace_at`): +5 Gold.
  4. the Great Prophet, one per player (`_gp_capped`): the race, the patronage
     and the pass refuse a seat that has earned one.
  5. a Holy Site's healing faith counts the Dar-e Mehr's era Faith and none of
     a pillaged building's.
  6. the Solar Farm refuses a feature plot; the builder's water jobs read the
     owning city's governor (`_imp_gov_ok`).
  7. a spent Apostle neither evangelizes nor launches the Inquisition.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

_HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(_HERE.parent.parent / "gpu"))
sys.path.insert(0, str(_HERE))
from core import BatchSim, load_rules, load_fixture, fixture_paths, FIXTURES
from warmup import settle_all
import religion2_test as r2

ROW = 0


def fresh(rules, path) -> BatchSim:
    return settle_all(BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64))


def test_purchase(sim, rj) -> None:
    b = [x["id"] for x in rj["buildings"]].index("GRANARY")
    assert float(sim.rules_dev.b_cost[b]) == 32.0
    assert float(sim._b_cols(ROW)["buyCost"][0, b]) == 32.5
    price = sim._gold_price(ROW, sim._b_cols(ROW)["buyCost"][:, b] * sim.rules.gold_purchase_mult)
    assert float(price[0]) == 130.0, float(price[0])
    u = [x["id"] for x in rj["units"]].index("SLINGER")
    uprice = sim._gold_price(ROW, sim._type_cost[u:u + 1].double() * sim.rules.gold_purchase_mult)
    assert float(uprice[0]) == 65.0, float(uprice[0])
    print("  1 the Granary buys for 130, the Slinger for 65 OK")


def test_policy_unlock(sim, rj) -> None:
    gov = {g["id"]: i for i, g in enumerate(rj["governments"])}
    civ = {c["id"]: i for i, c in enumerate(rj["civics"])}
    snap = sim.snapshot()
    active = torch.ones(sim.B, dtype=torch.bool)
    sim.seat_ext[:, ROW] = True
    sim.civ_civics[:, ROW] = False
    for c in ("CODE_OF_LAWS", "POLITICAL_PHILOSOPHY"):
        sim.civ_civics[:, ROW, civ[c]] = True
    sim.civ_gov_held[:, ROW] = 0
    sim.civ_gov_chosen[:, ROW] = -1
    sim.civ_policies[:, ROW] = False
    sim._eff_version += 1
    turn0 = sim.turn

    def record(g: int) -> None:
        sim.apply_seat_actions(ROW, government=torch.full((sim.B,), g, dtype=torch.long))
        sim._seat_record_apply(ROW, active)

    def now() -> int:
        g, has = sim._adopted_gov(ROW)
        return int(g[0]) if bool(has[0]) else -1

    mx, drop, mn = sim.rules.civic_unlock
    assert (mx, drop, mn) == (50.0, 5.0, 10.0)
    sim.civ_civic_turn[:, ROW] = 10
    for t, want in ((11, 0.0), (12, mx), (13, mx - drop), (100, mn)):
        sim.turn = t
        assert float(sim._policy_unlock_cost(ROW)[0]) == want, (t, float(sim._policy_unlock_cost(ROW)[0]))
    # three turns past the window: 40, refused one short of it
    sim.civ_civic_turn[:, ROW] = 1
    sim.turn = 5
    cost = float(sim._policy_unlock_cost(ROW)[0])
    assert cost == mx - 2 * drop
    sim.civ_treasury[:, ROW] = cost - 1
    record(gov["OLIGARCHY"])
    assert now() == gov["AUTOCRACY"] and float(sim.civ_treasury[0, ROW]) == cost - 1
    sim.civ_treasury[:, ROW] = cost + 7
    record(gov["OLIGARCHY"])
    assert now() == gov["OLIGARCHY"] and float(sim.civ_treasury[0, ROW]) == 7.0, float(sim.civ_treasury[0, ROW])
    # the same government again changes nothing and costs nothing
    record(gov["OLIGARCHY"])
    assert float(sim.civ_treasury[0, ROW]) == 7.0
    # in the window a change is free
    sim.civ_civic_turn[:, ROW] = 4
    sim.civ_treasury[:, ROW] = 0.0
    record(gov["CLASSICAL_REPUBLIC"])
    assert now() == gov["CLASSICAL_REPUBLIC"] and float(sim.civ_treasury[0, ROW]) == 0.0
    sim.turn = turn0
    sim.restore(snap)
    print("  2 the policy unlock: free, 50 dropping 5 to 10, paid once, refused short OK")


def test_minor_palace(sim) -> None:
    alive = sim.citystate_alive[0]
    assert bool(alive.any()), "the world has no living city-state"
    assert not bool(sim._palace_at(sim.FREE_ROW, slice(0, sim.RC)).any())
    real = sim._palace_at

    def gold_of(row: int, palace: bool) -> float:
        sim._palace_at = real if palace else (lambda r, sl: sim.city_is_cap[:, r, sl])
        sim._eff_version += 1
        g = float(sim._seat_city_stats(row)[0][0, 0, 2])
        sim._palace_at = real
        sim._eff_version += 1
        return g

    # the first minor whose city earns no Gold of its own without the Palace:
    # taking the Palace away also takes 10% off the city's other yields, so a
    # plot paying Gold would blur the flat +5 under test
    s = next((m for m in alive.nonzero().flatten().tolist()
              if gold_of(sim._CITY_MINOR0 + m, False) == 0.0), -1)
    assert s >= 0, "every city-state's city earns Gold of its own"
    row = sim._CITY_MINOR0 + s
    assert bool(sim._palace_at(row, slice(0, sim.RC))[0, 0])
    gold, bare = gold_of(row, True), gold_of(row, False)
    assert gold - bare == 5.0, (gold, bare)
    print(f"  3 a city-state's Palace pays +5 Gold (minor {s}: {bare} -> {gold}) OK")


def test_prophet_cap(sim) -> None:
    snap = sim.snapshot()
    p = sim._prophet_cls
    active = torch.ones(sim.B, dtype=torch.bool)
    assert not bool(sim._gp_capped(ROW, p)[0])
    sim.civ_gp_earned[:, ROW, p] = 1
    assert bool(sim._gp_capped(ROW, p)[0]) and not bool(sim._gp_capped(ROW, 0 if p else 1)[0])
    sim.civ_gpp[:, ROW, p] = 1.0e6
    sim.civ_faith[:, ROW] = 1.0e6
    sim.civ_treasury[:, ROW] = 1.0e6
    earned = int(sim.gp_earned[0, p])
    sim._advance_great_people(ROW, active)
    assert int(sim.civ_gp_earned[0, ROW, p]) == 1 and int(sim.gp_earned[0, p]) == earned
    assert float(sim.civ_gpp[0, ROW, p]) >= 1.0e6, "the points should wait"
    done = sim._patronize(ROW, active, torch.full((sim.B,), p, dtype=torch.long), gold=False)
    assert not bool(done[0]), "a capped seat patronized a Great Prophet"
    sim.restore(snap)
    print("  4 the Great Prophet, one per player: the race and patronage refuse OK")


def test_holy_site_faith(rules, rj, path) -> None:
    sim = r2.build(rules, path)
    r, g, j = 0, 1, 0
    r2.isolate_faith(sim, r)
    _wipe = sim.district[0] == sim._hs_idx
    sim.district[0, _wipe] = -1
    hs = r2.make_holy_site(sim, r, j)
    sim.tile_seat[0, hs] = g
    sim.tile_city[0, hs] = int(sim.city_id[0, g, j])
    sim._tile_owner_ver += 1
    bids = [x["id"] for x in rj["buildings"]]
    dm = bids.index("DAR_E_MEHR")
    sim.city_bldg[:, g, :, sim._shrine_bidx] = False
    sim.city_bldg[0, g, j, sim._shrine_bidx] = True
    sim.city_bldg[0, g, j, dm] = True
    k = int(sim._bpe_col[dm])
    assert k >= 0, "the Dar-e Mehr carries no per-era row"
    era = sim._game_era()
    sim.city_bldg_era[0, g, j, k] = era
    sim._eff_version += 1
    now = int(sim._holy_site_faith()[0, hs])
    turn0 = sim.turn
    sim.turn = (era + 2) * sim._era_len  # two game eras on
    assert int(sim._holy_site_faith()[0, hs]) == now + 2, (now, int(sim._holy_site_faith()[0, hs]))
    sim.turn = turn0
    dm_faith = int(rj["buildings"][dm]["yields"][5])
    sim.city_bldg_pillaged[0, g, j, dm] = True
    sim._eff_version += 1
    assert int(sim._holy_site_faith()[0, hs]) == now - dm_faith, "a pillaged Dar-e Mehr still paid"
    print(f"  5 the Holy Site's faith: {now} -> {now + 2} two eras on, {now - dm_faith} pillaged OK")


def test_feature_and_jobs(sim) -> None:
    ids = sim._imp_ids
    solar = ids.index("SOLAR_FARM")
    live = (sim.feat_id >= 0) & ~sim.feat_stripped
    if sim._soil_fid >= 0:
        live = live & (sim.feat_id != sim._soil_fid)
    assert bool(live.any())
    assert not bool((sim._imp_ground_ok(solar) & live).any()), "a Solar Farm on a feature plot"
    # the builder's water jobs read the owning city's governor
    fish = ids.index("FISHERY")
    coast = torch.nonzero(sim.water[0] & sim.wpass[0] & (sim.feat_id[0] < 0) & (sim.res_imp[0] < 0)
                          & ~sim.tile_submerged[0] & ~sim.nwonder[0] & (sim.tile_seat[0] < 0)).flatten()
    t = next(int(x) for x in coast.tolist() if bool(sim._imp_ground_ok(fish)[0, int(x)]))
    snap = sim.snapshot()
    sim.tile_seat[0, t] = ROW
    sim._tile_owner_ver += 1
    _ut, _uc = int(sim._imp_unlock[fish]), int(sim._imp_unlock_civic[fish])
    if _ut >= 0:
        sim.civ_techs[:, ROW, _ut] = True
    if _uc >= 0:
        sim.civ_civics[:, ROW, _uc] = True
    base = bool(sim._seat_job_mask(ROW)[0, t])
    real = sim._imp_gov_ok
    sim._imp_gov_ok = lambda row, k: torch.ones(sim.B, sim.T, dtype=torch.bool)
    held = bool(sim._seat_job_mask(ROW)[0, t])
    sim._imp_gov_ok = real
    sim.restore(snap)
    assert not base and held, (base, held)
    print("  6 no Solar Farm on a feature; a Fishery job where the governor holds Aquaculture OK")


def main() -> int:
    rules = load_rules()
    rj = json.loads((FIXTURES / "rules.json").read_text())
    path = fixture_paths()[0]
    sim = fresh(rules, path)
    test_purchase(sim, rj)
    test_policy_unlock(sim, rj)
    test_minor_palace(sim)
    test_prophet_cap(sim)
    test_holy_site_faith(rules, rj, path)
    test_feature_and_jobs(fresh(rules, path))
    print("eco residue OK")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/cpu/city/eco-residue.test.ts
/**
 * THE ECONOMY RESIDUE, TS side: a building's purchase price off its
 * fractional scaled cost, the policy unlock's Gold, a city-state's Palace, the
 * Great Prophet's one-per-player cap, the Dar-e Mehr's era Faith in a Holy
 * Site's healing faith, the renewable and water rows' feature refusal and the
 * builder's governor-gated jobs. The GPU twin is tests/gpu/eco_residue_test.py.
 */
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, settleAt, tileAtCoords, grantCivics, grantTechs, standDistrict } from '../helpers';
import { emptySeat, seatOf, seatOfCityState, setTileOwner } from '../../../cpu/core/seats';
import { buildingPurchaseCost, buildingFaithCost, unitPurchaseCost } from '../../../cpu/core/game';
import { applySeatActionRecord } from '../../../cpu/core/phase';
import { faithPrice, goldPrice, makeYieldCtx, policyUnlockCost, seatGovernment } from '../../../cpu/core/effects';
import { minorCity } from '../../../cpu/core/cityStates';
import { computeCityStats } from '../../../cpu/core/city';
import { advanceGreatPeople, gpCapped, patronizeGreatPerson } from '../../../cpu/core/greatPeople';
import { holySiteFaith } from '../../../cpu/core/units';
import { gameEraIndex } from '../../../cpu/core/yields';
import { validImprovements, validImprovementsIn } from '../../../cpu/core/rules';
import { governorsOf } from '../../../cpu/core/governors';
import { GOVERNOR_INDEX, GOVERNOR_PROMOTION_INDEX, promotionBitValue } from '../../../cpu/data/governors';
import { builderJobAt, jobCtx } from '../../../cpu/core/targetSites';
import { BUILDINGS } from '../../../cpu/data/buildings';
import { GREAT_PEOPLE, GP_CLASSES } from '../../../cpu/data/greatPeople';
import { GOVERNMENT_LIST, POLICY_LIST } from '../../../cpu/data/policies';
import { CIVIC_UNLOCK_MAX_COST, CIVIC_UNLOCK_MIN_COST, CIVIC_UNLOCK_PER_TURN_DROP, GAME_SPEED, POLICY_UNLOCK_K_BASE, POLICY_UNLOCK_K_PER_TECH, POLICY_UNLOCK_ROUND } from '../../../cpu/data/constants';
import { tilesWithin } from '../../../world/hex';
import type { CityState, GameState, SeatActionRecord } from '../../../cpu/core/types';

const GOV = (id: string) => GOVERNMENT_LIST.findIndex((g) => g.id === id);
const REC = (government?: number, policies?: number[]): SeatActionRecord =>
  ({ production: [], tech: null, civic: null, units: [], government, policies });

describe('a building buys off its fractional scaled cost', () => {
  it('prices the Granary (Cost 65) at 130 gold and 65 faith, as the lab record reads', () => {
    const state = makeState();
    settleAt(state, tileAtCoords(state.map, 5, 5).index);
    expect(BUILDINGS.GRANARY.cost).toBe(32);
    expect(BUILDINGS.GRANARY.buyCost).toBe(65 * GAME_SPEED);
    expect(goldPrice(state, 0, buildingPurchaseCost(state, 0, 'GRANARY'))).toBe(130);
    expect(faithPrice(state, 0, buildingFaithCost(state, 0, 'GRANARY'))).toBe(65);
    expect(goldPrice(state, 0, buildingPurchaseCost(state, 0, 'WORKSHOP'))).toBe(390);
    // a unit keeps the truncated cost: the Slinger (Cost 35) buys for 65
    expect(goldPrice(state, 0, unitPurchaseCost(state, 'SLINGER', 0))).toBe(65);
  });
});

describe('the policy unlock', () => {
  function scene(): GameState {
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    settleAt(state, tileAtCoords(state.map, 5, 5).index, 0);
    grantCivics(state, 'CODE_OF_LAWS', 'POLITICAL_PHILOSOPHY');
    return state;
  }

  it('is free the turn after a civic, then the maximum, dropping a step a turn to the minimum, times k', () => {
    const state = scene();
    const s = seatOf(state, 0)!;
    expect([CIVIC_UNLOCK_MAX_COST, CIVIC_UNLOCK_PER_TURN_DROP, CIVIC_UNLOCK_MIN_COST]).toEqual([50, 5, 10]);
    expect([POLICY_UNLOCK_K_BASE, POLICY_UNLOCK_K_PER_TECH, POLICY_UNLOCK_ROUND]).toEqual([15, 1, 5]);
    s.government.civicTurn = 10;
    // no techs: k = 1.5 — 75, 67.5 rounded half up to 70, the floor 15
    s.research.techs = [];
    const at = (turn: number) => { state.turn = turn; return policyUnlockCost(state, 0); };
    expect([at(11), at(12), at(13), at(100)]).toEqual([0, 75, 70, 15]);
    // 31 techs: k = 4.6 — the lab's jump to 230 at 31 techs, 207 to 205, the
    // floor 46 to 45
    s.research.techs = Array.from({ length: 31 }, (_, i) => `TECH_${i}`);
    expect([at(11), at(12), at(13), at(100)]).toEqual([0, 230, 205, 45]);
  });

  it('charges a government change outside the window once, and refuses one the purse cannot meet', () => {
    const state = scene();
    const s = seatOf(state, 0)!;
    s.government.civicTurn = 1;
    s.research.techs = [];
    state.turn = 5; // three turns past the window: the row's 40 at k = 1.5
    const cost = policyUnlockCost(state, 0);
    expect(cost).toBe(60);
    s.treasury = cost - 1;
    applySeatActionRecord(state, s, REC(GOV('OLIGARCHY')));
    expect(seatGovernment(state, 0)).toBe('AUTOCRACY');
    expect(s.treasury).toBe(cost - 1);
    s.treasury = cost + 7;
    // the government and a new card set in one record pay once
    const card = POLICY_LIST.findIndex((p) => p.id === 'DISCIPLINE');
    applySeatActionRecord(state, s, REC(GOV('OLIGARCHY'), [card]));
    expect(seatGovernment(state, 0)).toBe('OLIGARCHY');
    expect(s.government.policies).toContain('DISCIPLINE');
    expect(s.treasury).toBe(7);
    // naming what the seat already holds changes nothing and costs nothing
    applySeatActionRecord(state, s, REC(GOV('OLIGARCHY'), [card]));
    expect(s.treasury).toBe(7);
  });

  it('is free in the window', () => {
    const state = scene();
    const s = seatOf(state, 0)!;
    s.government.civicTurn = 4;
    state.turn = 5;
    s.treasury = 0;
    applySeatActionRecord(state, s, REC(GOV('CLASSICAL_REPUBLIC')));
    expect(seatGovernment(state, 0)).toBe('CLASSICAL_REPUBLIC');
    expect(s.treasury).toBe(0);
  });
});

describe('a city-state holds the Palace', () => {
  it('its city earns the Palace\'s +5 Gold and its other yields', () => {
    const state = makeState(makeMap(14, 14, 'GRASSLAND'));
    const center = tileAtCoords(state.map, 6, 6);
    const cs: CityState = {
      ...emptySeat(seatOfCityState(0)),
      id: 0, name: 'CS0', type: 'religious', centerIndex: center.index, population: 1, envoys: {}, met: [0],
    };
    for (const t of tilesWithin(state.map, 6, 6, 1)) setTileOwner(t, seatOfCityState(0));
    state.cityStates.push(cs);
    state.cityStateMax = 1;
    const city = minorCity(cs);
    expect(city.buildings).toContain('PALACE');
    const withPalace = computeCityStats(state, city).total;
    const bare = computeCityStats(state, { ...city, buildings: city.buildings.filter((b) => b !== 'PALACE') }).total;
    expect(withPalace.gold - bare.gold).toBe(BUILDINGS.PALACE.yields!.gold);
    expect(BUILDINGS.PALACE.yields!.gold).toBe(5);
    expect(withPalace.production).toBeGreaterThan(bare.production);
  });
});

describe('the Great Prophet, one per player', () => {
  it('a seat that has earned its Prophet takes no other', () => {
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    settleAt(state, tileAtCoords(state.map, 5, 5).index, 0);
    const s = seatOf(state, 0)!;
    const ci = GP_CLASSES.indexOf('PROPHET');
    expect(gpCapped(s, 'PROPHET')).toBe(false);
    s.gpEarned.push(GREAT_PEOPLE.PROPHET[0].id);
    expect(gpCapped(s, 'PROPHET')).toBe(true);
    expect(gpCapped(s, 'GENERAL')).toBe(false);
    s.gpp.PROPHET = 100000;
    s.faith = 100000;
    s.treasury = 100000;
    const before = state.units.length;
    advanceGreatPeople(state, 0);
    expect(state.units.filter((u) => u.type === 'PROPHET').length).toBe(0);
    expect(state.units.length).toBe(before);
    expect(s.gpp.PROPHET).toBeGreaterThanOrEqual(100000); // the points wait
    expect(patronizeGreatPerson(state, 0, ci, 'faith').ok).toBe(false);
  });
});

describe('a Holy Site\'s healing faith counts the Dar-e Mehr\'s era Faith', () => {
  it('adds a Faith per era since the Dar-e Mehr was built', () => {
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    const city = settleAt(state, tileAtCoords(state.map, 5, 5).index, 0);
    grantTechs(state, 'ASTROLOGY');
    const hs = tileAtCoords(state.map, 6, 5);
    standDistrict(state, city, 'HOLY_SITE', hs.index);
    city.buildings.push('SHRINE', 'TEMPLE', 'DAR_E_MEHR');
    const ctx = makeYieldCtx(state, 0);
    const era = gameEraIndex(state);
    city.buildingEras = { DAR_E_MEHR: era };
    const now = holySiteFaith(state, hs, ctx);
    city.buildingEras = { DAR_E_MEHR: era - 2 };
    expect(holySiteFaith(state, hs, ctx)).toBe(now + 2 * BUILDINGS.DAR_E_MEHR.yieldsPerEra!.faith!);
    // a pillaged building pays none of its Faith
    city.pillagedBuildings = ['DAR_E_MEHR'];
    expect(holySiteFaith(state, hs, ctx)).toBe(now - BUILDINGS.DAR_E_MEHR.yields!.faith!);
  });
});

describe('a row with no Improvement_ValidFeatures row refuses a feature plot', () => {
  it('the Solar Farm refuses Woods; the Fishery and the Polder refuse a Reef', () => {
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    const t = tileAtCoords(state.map, 3, 3);
    const opts = { unlocks: null, ownsTile: () => true, map: state.map };
    expect(validImprovementsIn(t, opts)).toContain('SOLAR_FARM');
    t.feature = 'WOODS';
    expect(validImprovementsIn(t, opts)).not.toContain('SOLAR_FARM');
    const h = tileAtCoords(state.map, 3, 6);
    h.elevation = 'HILLS';
    expect(validImprovementsIn(h, opts)).toContain('WIND_FARM');
    h.feature = 'WOODS';
    expect(validImprovementsIn(h, opts)).not.toContain('WIND_FARM');
    const c = tileAtCoords(state.map, 8, 8);
    c.terrain = 'COAST';
    const water = { ...opts, civ: 'NETHERLANDS', govPromos: new Set(['AQUACULTURE']) };
    const bare = validImprovementsIn(c, water);
    expect(bare).toEqual(expect.arrayContaining(['FISHERY', 'POLDER', 'OFFSHORE_WIND_FARM']));
    c.feature = 'REEF';
    const onReef = validImprovementsIn(c, water);
    expect(onReef).not.toContain('FISHERY');
    expect(onReef).not.toContain('POLDER');
    expect(onReef).not.toContain('OFFSHORE_WIND_FARM');
  });
});

describe('the builder\'s job plane reads the owning city\'s governor', () => {
  it('a City Park is a job only where the governor holds its promotion', () => {
    const state = makeState(makeMap(12, 12, 'GRASSLAND'));
    const city = settleAt(state, tileAtCoords(state.map, 5, 5).index, 0);
    grantCivics(state, 'GAMES_AND_RECREATION');
    // a Snow plot: no Farm, Mine or Lumber Mill, so only the Park could work it
    const t = tilesWithin(state.map, 5, 5, 1).find((x) => x.index !== city.centerIndex)!;
    t.terrain = 'SNOW';
    t.feature = null;
    t.resource = null;
    setTileOwner(t, 0, city.id);
    expect(builderJobAt(state, jobCtx(state, 0), t)).toBe(false);
    const g = governorsOf(seatOf(state, 0)!)[GOVERNOR_INDEX.LIANG];
    g.appointed = true;
    g.cityId = city.id;
    g.establishTurns = 0;
    g.promotions = promotionBitValue(GOVERNOR_PROMOTION_INDEX.PARKS_AND_RECREATION!);
    expect(validImprovements(state, t, 0)).toContain('CITY_PARK');
    expect(builderJobAt(state, jobCtx(state, 0), t)).toBe(true);
  });
});
